src/python/swigtest.py

```diff
@@ -0,0 +1,201 @@
+
+    return d
+
+def avrpart_to_mem(avrpart):
+
+    if str(type(avrpart)).find('AVRPART') < 0:
+        raise Exception(f"wrong argument: {type(avrpart)}, expecting swig_avrdude.AVRPART")
+
+    res = []
+    m = ad.lfirst(avrpart.mem)
+    while m:
+        mm = ad.ldata_avrmem((m))
+        res.append(avrmem_to_dict(mm))
+        m = ad.lnext(m)
+
+    return res
+
+def getavr(name: str):
+
+    p = ad.locate_part(ad.cvar.part_list, name)
+    if not p:
+        print(f"No part named {name} found")
+        return
+    pp = avrpart_to_dict(p)
+    mm = avrpart_to_mem(p)
+    print(f"AVR part {name} found as {pp['desc']}, or {pp['id']}")
+    print(f"Definition in {pp['config_file']}, line {pp['lineno']}")
+
+    print("")
+    print("Memory overview:")
+    print( "Name        size   paged   page_size num_pages")
+    for m in mm:
+        print(f"{m['desc']:11s} {m['size']:6d}  {str(m['paged']):5s}   {m['page_size']:4d}      {m['num_pages']:3d}")
+
+    print("")
+    print("Device variants:")
+    v = ad.lfirst(p.variants)
+    while v:
+        vv = ad.ldata_string(v)
+        print(vv)
+        v = ad.lnext(v)
+    print("")
+
+def start_programmer(programmer: str, port: str, avrpart):
+    g = ad.locate_programmer(ad.cvar.programmers, programmer)
+    g.initpgm()
+    g.setup()
+    g.open(port)
+    g.enable(p)
+    g.initialize(p)
+
+    return g
+
+def stop_programmer(g):
+    g.disable()
+    g.close()
+    g.teardown()
+
+def message_type(msglvl: int):
+    tnames = ('OS error', 'error', 'warning', 'info', 'notice',
+              'notice2', 'debug', 'trace', 'trace2')
+    msglvl -= ad.MSG_EXT_ERROR # rebase to 0
+    if msglvl > len(tnames):
+        return 'unknown msglvl'
+    else:
+        return tnames[msglvl]
+
+# rough equivalent of avrdude_message2()
+# first argument is either "stdout" or "stderr"
+#
+# install callback with ad.set_msg_callback(msg_callback)
+def msg_callback(target: str, lno: int, fname: str, func: str,
+                 msgmode: int, msglvl: int, msg: str):
+    if ad.cvar.verbose >= msglvl:
+        s = ""
+        if msgmode & ad.MSG2_PROGNAME:
+            s += ad.cvar.progname
+            if ad.cvar.verbose >= ad.MSG_NOTICE and (msgmode & ad.MSG2_FUNCTION) != 0:
+                s += " " + func + "()"
+            if ad.cvar.verbose >= ad.MSG_DEBUG and (msgmode & ad.MSG2_FILELINE) != 0:
+                n = os.path.basename(fname)
+                s += f" [{n}:{lno}]"
+            if (msgmode & ad.MSG2_TYPE) != 0:
+                s += " " + message_type(msglvl)
+            s += ": "
+        elif (msgmode & ad.MSG2_INDENT1) != 0:
+            s = (len(ad.cvar.progname) + 1) * ' '
+        elif (msgmode & ad.MSG2_INDENT2) != 0:
+            s = (len(ad.cvar.progname) + 2) * ' '
+        s += msg
+        try:
+            print(s, end='')
+        except UnicodeDecodeError:
+            print("message contained invalid characters")
+
+# very simplified progress callback
```

Comment:
Do you want to bring that up to speed now that PR #1724 documents how `update_progress()` should *properly* work>

Reply:
> Do you want to bring that up to speed now that PR #1724 documents how `update_progress()` should _properly_ work>

I'll have a look at it, but keep in mind: `swigtest.py` is rather meant as a test vehicle, to allow some interactive Python operation for testing the SWIG wrapper has everything that is needed.

My main goal now is to demonstrate that a GUI wrapper around the library offers a lot more than wrapping a GUI around the CLI.